README.md
# Apollo Engine

[![npm version](https://badge.fury.io/js/apollo-engine.svg)](https://badge.fury.io/js/apollo-engine)
[![Build Status](https://circleci.com/gh/apollographql/apollo-engine-js.svg?style=svg)](https://circleci.com/gh/apollographql/apollo-engine-js)

This package integrates the Apollo Engine Proxy with your GraphQL server.

Read the docs at https://apollographql.com/docs/engine/setup-node.html

Release notes for this package are at https://www.apollographql.com/docs/engine/proxy-release-notes.html

For feature requests, bug reports, or general questions or feedback on Apollo Engine Proxy, please use [this form](https://engine.apollographql.com/login?overlay=SupportRequestNoAccount) instead of opening an issue on this repository.

Because this repo contains a wrapper around the Apollo Engine Proxy binary, feature requests and support cases for the underlying proxy will be closed. Please only open an issue in this repo if you believe there is an issue with the wrapper/middleware code or if you would like a feature represented in the wrapper itself.

For all other requests or questions, we encourage you to [use this form](https://engine.apollographql.com/login?overlay=SupportRequestNoAccount). The #engine channel in the [Apollo Slack](apollographql.com/#slack) can also be a great resource for general questions about Apollo Engine. Following these guidelines will ensure that your requests are seen and addressed as quickly as possible and also allows us to better collect and iterate on feedback.


CHANGELOG.md
Release notes for this package are at https://www.apollographql.com/docs/engine/proxy/release-notes.html


src/__tests__/child.js
// This script is run via child_process.fork by tests in order to test engine's
// process shutdown behavior. This file must be kept Node 4-compatible.

// Note that you may need to manually run "npm install"/"tsc" before running
// tests to pick up changes to the non-test Engine code.
const lib = require('../../lib');
const http = require('http');

const config = {
  apiKey: 'dummy',
  reporting: {
    disabled: true,
  },
  origins: [
    {
      http: {
        url: 'http://127.0.0.1:1234/graphql',
      },
    },
  ],
};

function maybeDie() {
  if (process.env.AEJ_TEST_UNCAUGHT_EXCEPTION) {
    throw new Error('bam');
  }
  if (process.env.AEJ_TEST_PROCESS_EXIT) {
    process.exit(42);
  }
}

if (process.env.AEJ_TEST_LAUNCHER) {
  const launcher = new lib.ApolloEngineLauncher(config);

  // Make sure thrown errors throw rather than end up in Promises.  (This is
  // isn't just inlined into the catch call later because otherwise prettier's
  // formatting will add a trailing comma after the function definition, which
  // breaks running tests on Node 6 and older.)
  function throwSoon(err) {
    process.nextTick(() => {
      throw err;
    });
  }

  launcher
    .start()
    .then(() => {
      process.send({ pid: launcher.child.pid });
      setTimeout(maybeDie, 2);
    })
    .catch(throwSoon);
} else {
  const engine = new lib.ApolloEngine(config);

  engine.listen({ port: 0, httpServer: http.createServer() }, () => {
    process.send({ pid: engine.launcher.child.pid });
    setTimeout(maybeDie, 2);
  });
}


src/__tests__/engine.test.ts
import { graphqlExpress, graphqlConnect } from 'apollo-server-express';
import * as bodyParser from 'body-parser';
import * as connect from 'connect';
import { NextHandleFunction } from 'connect';
import * as express from 'express';
import * as http from 'http';
import * as qs from 'qs';
import * as urlModule from 'url';

import {
  schema,
  rootValue,
  verifyEndpointSuccess,
  verifyEndpointGet,
  verifyEndpointError,
  verifyEndpointFailure,
  verifyEndpointBatch,
} from './schema';
import { processIsRunning, devNull } from './util';
import {
  runSuite,
  runSuitesForHttpServerFramework,
  runCleanupTests,
} from './engine-common';

import { ApolloEngine } from '../engine';
import * as os from 'os';

runSuitesForHttpServerFramework('express', {
  createApp() {
    const path = '/graphql';
    const app = express();
    app.get(`${path}/ping`, (req, res) => {
      res.json({ pong: true });
    });
    app.use(
      path,
      bodyParser.json(),
      graphqlExpress({
        schema,
        rootValue,
        tracing: true,
        cacheControl: true,
      }),
    );
    return app;
  },
  serverForApp(app: any) {
    return http.createServer(app);
  },
  appParameter: 'expressApp',
});

function connectQuery(
  req: http.IncomingMessage,
  res: http.ServerResponse,
  next: Function,
) {
  const parsedUrl = urlModule.parse(req.url!);
  (req as any).query = qs.parse(parsedUrl.query!);
  next();
}

runSuitesForHttpServerFramework('connect', {
  createApp() {
    const path = '/graphql';
    const app = connect().use(connectQuery);
    app.use(
      `${path}/ping`,
      (req: http.IncomingMessage, res: http.ServerResponse) => {
        res.end(JSON.stringify({ pong: true }));
      },
    );
    app.use(path, bodyParser.json() as NextHandleFunction);
    app.use(path, graphqlConnect({
      schema,
      rootValue,
      tracing: true,
      cacheControl: true,
    }) as NextHandleFunction);
    return app;
  },
  serverForApp(app: any) {
    return http.createServer(app);
  },
  appParameter: 'connectApp',
});

test('can pass a string as a port', async () => {
  const httpServer = http.createServer();
  const engine = new ApolloEngine({
    apiKey: 'faked',
    logging: {
      level: 'WARN',
      destination: 'STDERR',
    },
    reporting: {
      disabled: true,
    },
  });
  try {
    const p = new Promise(resolve =>
      engine.listen({ port: '0', httpServer }, resolve),
    );
    await p;
  } finally {
    await engine.stop();
    httpServer.close();
  }
});

test('must specify apiKey', async () => {
  const httpServer = http.createServer();
  const engine = new ApolloEngine();
  const p = new Promise((resolve, reject) => {
    engine.on('error', e => resolve(e));
    engine.listen(
      {
        port: '0',
        httpServer,
        launcherOptions: { proxyStderrStream: devNull() },
      },
      () => reject(new Error('should not start')),
    );
  });
  await p;
  httpServer.close();
});

describe('env var', () => {
  let oldValue: string | undefined;
  beforeEach(() => {
    oldValue = process.env.ENGINE_API_KEY;
  });
  afterEach(() => {
    if (oldValue === undefined) {
      delete process.env.ENGINE_API_KEY;
    } else {
      process.env.ENGINE_API_KEY = oldValue;
    }
  });

  test('can specify apiKey as env var', async () => {
    const httpServer = http.createServer();
    const engine = new ApolloEngine();
    process.env.ENGINE_API_KEY = 'faked';
    try {
      const p = new Promise(resolve =>
        engine.listen(
          {
            port: '0',
            httpServer,
            launcherOptions: { proxyStderrStream: devNull() },
          },
          resolve,
        ),
      );
      await p;
    } finally {
      await engine.stop();
      httpServer.close();
    }
  });
});

describe('launch failure', () => {
  let engine: ApolloEngine | null = null;
  let httpServer: http.Server | null = null;
  beforeEach(() => {
    engine = null;
    httpServer = null;
  });
  afterEach(async () => {
    if (engine !== null) {
      const child = engine['launcher']['child'];
      if (child) {
        await engine.stop();
        expect(processIsRunning(child.pid)).toBe(false);
      }
      engine = null;
    }

    if (httpServer) {
      httpServer.close();
    }
  });
  test('emits error on invalid config', async () => {
    engine = new ApolloEngine({
      apiKey: 'faked',
      logging: {
        level: 'INVALID',
      },
      reporting: {
        disabled: true,
      },
    });

    const start = +new Date();
    httpServer = http.createServer();
    const p = new Promise((resolve, reject) => {
      // Help TS understand that these variables are still set.
      httpServer = httpServer!;
      engine = engine!;
      // We expect to get an error, so that's why we're *resolving* with it.
      engine!.once('error', err => {
        resolve(err.message);
      });
      engine!.listen(
        {
          httpServer,
          port: 0,
          launcherOptions: { proxyStderrStream: devNull() },
        },
        () => reject(new Error('Engine should not listen successfully')),
      );
    });
    await expect(p).resolves.toMatch(
      /Engine crashed due to invalid configuration/,
    );
    const end = +new Date();
    expect(end - start).toBeLessThan(5000);
  });
  if (os.type() === 'Windows_NT') {
    // Named pipes should actually work
    test('using pipePath on Windows should actually work', async () => {
      httpServer = http.createServer();
      httpServer = httpServer!;
      engine = new ApolloEngine({
        apiKey: 'faked',
        logging: {
          level: 'DEBUG',
          destination: 'STDERR',
        },
        reporting: {
          disabled: true,
        },
      });
      try {
        const p = new Promise(resolve => {
          httpServer = httpServer!;
          engine!.listen(
            { pipePath: String.raw`\\.\pipe\foo`, httpServer },
            resolve,
          );
        });
        await p;
      } finally {
        await engine.stop();
        httpServer.close();
      }
    });

    test('using pipe in port variable on Windows should actually work', async () => {
      httpServer = http.createServer();
      httpServer = httpServer!;
      engine = new ApolloEngine({
        apiKey: 'faked',
        logging: {
          level: 'DEBUG',
          destination: 'STDERR',
        },
        reporting: {
          disabled: true,
        },
      });
      try {
        const p = new Promise(resolve => {
          httpServer = httpServer!;
          engine!.listen(
            { port: String.raw`\\.\pipe\foo`, httpServer },
            resolve,
          );
        });
        await p;
      } finally {
        await engine.stop();
        httpServer.close();
      }
    });
  } else {
    test('using pipePath on non-Windows results in a configuration error', async () => {
      engine = new ApolloEngine({
        apiKey: 'faked',
        reporting: {
          disabled: true,
        },
      });

      const start = +new Date();
      httpServer = http.createServer();
      const p = new Promise((resolve, reject) => {
        // Help TS understand that these variables are still set.
        httpServer = httpServer!;
        engine = engine!;
        // We expect to get an error, so that's why we're *resolving* with it.
        engine!.once('error', err => {
          resolve(err.message);
        });
        engine!.listen(
          {
            httpServer,
            pipePath: 'anything',
            launcherOptions: { proxyStderrStream: devNull() },
          },
          () => reject(new Error('Engine should not listen successfully')),
        );
      });
      await expect(p).resolves.toMatch(
        /Engine crashed due to invalid configuration/,
      );
      const end = +new Date();
      expect(end - start).toBeLessThan(5000);
    });
  }
});

runCleanupTests(false);


src/__tests__/engine-common.ts
import * as http from 'http';
import * as request from 'request';
import { join as pathJoin } from 'path';
import { fork } from 'child_process';
import { stub, SinonStub } from 'sinon';

import {
  schema,
  rootValue,
  verifyEndpointSuccess,
  verifyEndpointGet,
  verifyEndpointError,
  verifyEndpointFailure,
  verifyEndpointBatch,
} from './schema';
import { processIsRunning, devNull } from './util';

import { ApolloEngine } from '../engine';

const acceptableEndings = ['/', '?', '?123', '/?123'];

export function runSuite(
  before: Function,
  hasTracing: boolean,
  frameworkName: string,
) {
  let url: string;

  // micro has an unconfigurable behavior to console.error any error thrown by a
  // handler (https://github.com/zeit/micro/issues/329).  We use sinon to
  // override console.error; however, we use callThrough to ensure that by
  // default, it just calls console.error normally. The tests that throw errors
  // tell the stub to "stub out" console.error on the first call.
  let consoleErrorStub: SinonStub;

  beforeEach(async () => {
    consoleErrorStub = stub(console, 'error');
    consoleErrorStub.callThrough();

    url = await before();
  });

  afterEach(() => {
    consoleErrorStub.restore();
  });

  test('processes successful query', () => {
    return verifyEndpointSuccess(url, hasTracing);
  });
  acceptableEndings.forEach(acceptableEnding => {
    test(`using server endpoint ${acceptableEnding}`, () => {
      return verifyEndpointSuccess(url + acceptableEnding, hasTracing);
    });
  });
  test('processes successful GET query', () => {
    return verifyEndpointGet(url, hasTracing);
  });
  test('processes invalid query', () => {
    if (frameworkName === 'micro') {
      consoleErrorStub.onFirstCall().returns(undefined);
    }
    return verifyEndpointFailure(url);
  });
  test('processes query that errors', () => {
    return verifyEndpointError(url);
  });
  test('processes batched queries', () => {
    return verifyEndpointBatch(url, hasTracing);
  });
  test('returns cache information', async () => {
    const body: any = await verifyEndpointSuccess(url, hasTracing);
    expect(
      body['extensions'] && body['extensions']['cacheControl'],
    ).toBeDefined();
  });

  test('http proxying works', done => {
    const childUrl = `${url}/ping`;
    request(childUrl, (err, response, body) => {
      expect(err).toBe(null);
      expect(body).toBe('{"pong":true}');
      done();
    });
  });
}

export function runSuitesForHttpServerFramework(
  frameworkName: string,
  { createApp, serverForApp, appParameter }: any,
) {
  describe(`${frameworkName} integration`, () => {
    let httpServers: http.Server[] = [];
    let engine: ApolloEngine | null;

    beforeEach(() => {
      engine = null;
      httpServers = [];
    });
    afterEach(async () => {
      if (engine) {
        await engine.stop();
      }
      httpServers.forEach(server => server.close());
    });

    function gqlServer() {
      const app = createApp();
      const server = serverForApp(app);
      httpServers.push(server);
      return server.listen().address().port;
    }

    describe('without engine', () => {
      runSuite(
        async () => {
          return `http://localhost:${gqlServer()}/graphql`;
        },
        true,
        frameworkName,
      );
    });

    describe('with engine', () => {
      runSuite(
        async () => {
          const app = createApp();
          engine = new ApolloEngine({
            apiKey: 'faked',
            logging: {
              level: 'WARN',
              destination: 'STDERR',
            },
            reporting: {
              disabled: true,
            },
            frontends: [
              {
                extensions: {
                  strip: ['tracing'], // ... but not cache control!
                },
              },
            ],
          });
          const p = new Promise(resolve => {
            engine!.listen(
              {
                // Let engineproxy get an ephemeral port; we'll learn about it in the
                // listening callback.
                port: 0,
                [appParameter]: app,
                launcherOptions: {
                  extraEnv: {
                    // engineproxy should only try to connect to our origin
                    // (reporting is diabled), which is on localhost, so this
                    // bad proxy should be ignored. This is a regression test
                    // from back when we would put unspecified IPs in the origin
                    // URL by default (instead of listening on 127.0.0.1 by
                    // default) and the Go http library would try to use the
                    // proxy (https://github.com/golang/go/issues/24737).
                    HTTP_PROXY: 'http://bad.proxy.example.com/',
                  },
                },
              },
              () => {
                resolve(`${engine!.engineListeningAddress.url}/graphql`);
              },
            );
          });
          return await p;
        },
        false,
        frameworkName,
      );
    });
  });
}

export function runCleanupTests(forLauncher: boolean) {
  describe('engineproxy cleaned up', () => {
    ['SIGINT', 'SIGTERM', 'SIGUSR2', 'uncaughtException', 'exit'].forEach(
      event => {
        test(`on ${event}`, async () => {
          // There is no SIGUSR2 on Windows.
          if (event === 'SIGUSR2' && process.platform === 'win32') {
            return;
          }

          const env: NodeJS.ProcessEnv = {
            ...process.env,
          };
          if (forLauncher) {
            env.AEJ_TEST_LAUNCHER = 't';
          }
          if (event === 'uncaughtException') {
            env.AEJ_TEST_UNCAUGHT_EXCEPTION = 't';
          }
          if (event === 'exit') {
            env.AEJ_TEST_PROCESS_EXIT = 't';
          }
          const child = fork(pathJoin(__dirname, 'child.js'), [], {
            env,
            // You may want to remove the following line to debug failures in
            // these tests.
            silent: true,
          });
          const proxyPid = await new Promise<number>(resolve => {
            child.on('message', m => {
              resolve(m.pid);
            });
          });

          // Verify that the proxy exists.
          process.kill(proxyPid, 0);

          const childDone = new Promise(resolve => {
            child.on('exit', resolve);
          });

          if (event.startsWith('SIG')) {
            child.kill(event);
          }
          await childDone;

          // 'exit' and 'uncaughtException' don't actually wait for the proxy to
          // be gone, so sleep a bit to avoid races in tests.
          if (event === 'exit' || event === 'uncaughtException') {
            await new Promise(resolve => setTimeout(resolve, 10));
          }

          // Verify that the proxy is gone.
          expect(() => process.kill(proxyPid, 0)).toThrow();
        });
      },
    );
  });
}


src/launcher.ts
import { EventEmitter } from 'events';
import { ChildProcess, spawn } from 'child_process';

import {
  EngineConfig,
  LauncherOptions,
  ListeningAddress,
  TcpListeningAddress,
} from './types';

// An arbitrary function; suitable as an EventEmitter event handler.
type eventHandler = (...args: any[]) => any;

// ApolloEngineLauncher knows how to run an engineproxy binary, wait for it to
// be listening, learn on what address it is listening, and restart it when it
// crashes. It doesn't know how to automatically connect it to a GraphQL server
// in this process --- that's what ApolloEngine is for. It's used to implement
// ApolloEngine, and it's an alternative to the Docker container for folks who
// want to put engineproxy in front of a non-Node GraphQL server.
export class ApolloEngineLauncher extends EventEmitter {
  private config: EngineConfig;
  private binary: string;
  private child: ChildProcess | null;
  private eventStopper?: () => void;

  // The constructor takes the same argument as ApolloEngine: the underlying
  // engineproxy config file.  Note that unlike the ApolloEngine constructor,
  // the config is mandatory, because you need to at least define an origin URL.
  public constructor(config: EngineConfig) {
    super();

    this.config = config;
    this.child = null;
    switch (process.platform) {
      case 'darwin':
        this.binary = require.resolve(
          'apollo-engine-binary-darwin/engineproxy_darwin_amd64',
        );
        break;
      case 'linux':
        this.binary = require.resolve(
          'apollo-engine-binary-linux/engineproxy_linux_amd64',
        );
        break;
      case 'win32':
        this.binary = require.resolve(
          'apollo-engine-binary-windows/engineproxy_windows_amd64.exe',
        );
        break;
      default:
        throw new Error('Unsupported platform');
    }
  }

  // start takes the same options as the launcherOptions option to
  // `ApolloEngine.listen`. It runs engineproxy, returning a Promise that
  // resolves once engineproxy is listening, or rejects if startup fails
  // (including due to a timeout). It restarts engineproxy if it exits for any
  // reason other than invalid config (emitting 'restarting' as it does so).
  // The Promise resolves to a structure telling on what port engineproxy is
  // listening.
  public start(options: LauncherOptions = {}): Promise<ListeningAddress> {
    if (this.child) {
      throw new Error(
        'Only call start() on an ApolloEngineLauncher object once',
      );
    }

    const spawnChild = () => {
      // We want to read from engineproxy's special listening reporter fd 3
      // (which we tell it about with an env var). We let it write directly to
      // our stdout and stderr (unless the user passes in their own output
      // streams) so we don't spend CPU copying output around (and if we crash
      // for some reason, engineproxy's output still gets seen). We don't care
      // about engineproxy's stdin.
      //
      // We considered having stdout and stderr always wrapped with a prefix. We
      // used to do this before we switched to JSON but apparently it was slow:
      // https://github.com/apollographql/apollo-engine-js/pull/50#discussion_r153961664
      // Users can use proxyStd*Stream to do this themselves, and we can make it
      // easier if it's popular.
      const stdio = ['ignore', 'inherit', 'inherit', 'pipe'];

      // If we are provided writable streams, ask child_process to create a pipe
      // which we will pipe to them. (We could put the streams directly in
      // `stdio` but this only works for pipes based directly on files.)
      if (options.proxyStdoutStream) {
        stdio[1] = 'pipe';
      }
      if (options.proxyStderrStream) {
        stdio[2] = 'pipe';
      }

      const args: string[] = ['-listening-reporter-fd=3'];
      const env = Object.assign({}, process.env);
      if (typeof this.config === 'string') {
        // Load config from a file. engineproxy will watch the file for changes.
        args.push(`-config=${this.config}`);
      } else {
        args.push(`-config=env`);
        env.ENGINE_CONFIG = JSON.stringify(this.config);
      }

      // Add extra arguments; used by ApolloEngine and by tests.
      if (options.extraArgs) {
        options.extraArgs.forEach(a => args.push(a));
      }

      // Add extra environment variables. Used by tests.
      if (options.extraEnv) {
        Object.keys(options.extraEnv).forEach(k => {
          env[k] = options.extraEnv![k];
        });
      }

      const child = spawn(this.binary, args, { stdio, env });
      this.child = child;

      // Hook up custom logging streams, if provided. We don't end the streams
      // when the child ends, as we may run several restarting childs against
      // one set of output streams.
      if (options.proxyStdoutStream) {
        child.stdout.pipe(options.proxyStdoutStream, { end: false });
      }
      if (options.proxyStderrStream) {
        child.stderr.pipe(options.proxyStderrStream, { end: false });
      }

      let listeningAddress = '';
      child.stdio[3].on('data', chunk => {
        listeningAddress += chunk.toString();
      });
      child.stdio[3].on('end', () => {
        // If we read something, then it started. (If not, then this is probably
        // just end of process cleanup.)
        if (listeningAddress !== '') {
          // Notify that proxy has started. The object is of the form `{ip:
          // "127.0.0.1", port: 1234}`.
          const la: ListeningAddress = JSON.parse(listeningAddress);
          // Convert IPs which mean "any address" (IPv4 or IPv6) into localhost
          // corresponding loopback ip. Note that the url field we're setting is
          // primarily for consumption by our test suite. If this heuristic is
          // wrong for your use case, explicitly specify a frontend host (in the
          // `frontends.host` field in your engine config, or in the `host`
          // option to ApolloEngine.listen).
          if (ApolloEngineLauncher.isTcp(la)) {
            // we are listening via TCP
            let hostForUrl = la.ip;
            if (la.ip === '' || la.ip === '::') {
              hostForUrl = 'localhost';
            }
            la.url = `http://${joinHostPort(hostForUrl, la.port)}`;
          }
          this.emit('start', la);
        }
      });
      // Re-emit any errors from talking to engineproxy.
      // XXX Not super clear if this will happen in practice, but at least
      //     if it does, doing it this way will make it clear that the error
      //     is coming from Engine.
      child.stdio[3].on('error', err => this.emit('error', err));

      // Connect shutdown hooks:
      child.on('exit', (code, signal) => {
        if (!this.child) {
          // It's not an error if we've killed it (due to timeout or stop()).
          return;
        }
        if (code === 78) {
          this.emit(
            'error',
            new Error('Engine crashed due to invalid configuration.'),
          );
          return;
        }

        if (code != null) {
          this.emitRestarting(`Engine crashed unexpectedly with code: ${code}`);
        }
        if (signal != null) {
          this.emitRestarting(
            `Engine was killed unexpectedly by signal: ${signal}`,
          );
        }
        spawnChild();
      });
    };

    spawnChild();

    this.setUpStopEvents(
      options.processCleanupEvents || [
        'exit',
        'uncaughtException',
        'SIGINT',
        'SIGTERM',
        // We mostly care about SIGUSR2 because nodemon uses it.
        'SIGUSR2',
      ],
    );

    return new Promise((resolve, reject) => {
      let startupErrorHandler: (error: Error) => void;
      let cancelTimeout: NodeJS.Timer;
      if (options.startupTimeout === undefined || options.startupTimeout > 0) {
        cancelTimeout = setTimeout(() => {
          if (this.child) {
            this.child.kill('SIGKILL');
            this.child = null;
          }
          this.removeListener('error', startupErrorHandler);
          reject(Error('engineproxy timed out'));
        }, options.startupTimeout || 5000);
      }

      this.on('start', (listeningAddress: ListeningAddress) => {
        clearTimeout(cancelTimeout);
        this.removeListener('error', startupErrorHandler);
        resolve(listeningAddress);
      });

      // Errors during startup turn into rejections of the 'start'
      // Promise. Later errors are just emitted as 'error' events (as are
      // startup errors).
      startupErrorHandler = (error: Error) => {
        clearTimeout(cancelTimeout);
        this.child = null;
        reject(error);
      };
      this.once('error', startupErrorHandler);
    });
  }

  // Stops the process. The returned Promise resolves once the child has exited.
  public stop(): Promise<void> {
    if (this.child === null) {
      throw new Error('No engine instance running!');
    }
    if (this.eventStopper) {
      this.eventStopper();
    }
    const childRef = this.child;
    this.child = null;
    return new Promise(resolve => {
      childRef.on('exit', () => {
        resolve();
      });
      childRef.kill();
    });
  }

  private static isTcp(
    listeningAddress: ListeningAddress,
  ): listeningAddress is TcpListeningAddress {
    return (<TcpListeningAddress>listeningAddress).port !== undefined;
  }

  private emitRestarting(error: string) {
    if (!this.emit('restarting', new Error(error))) {
      // No listeners; default to console.error.
      console.error(error);
    }
  }

  private setUpStopEvent(event: string): eventHandler {
    const handler = (maybeError?: Error) => {
      let childPromise = Promise.resolve();
      // Kill the child if it's running.
      if (this.child) {
        // Note that in the case of the 'exit' event we're not going to
        // actually get a chance to wait for anything to happen, but at least
        // we'll fire off the signal.
        childPromise = this.stop();
      }

      // For uncaughtException, rethrow the error. This does slightly change the
      // printed error (but not its stack trace) and converts the exit status of
      // 1 into 7, but it's pretty close. We do this synchronously rather than
      // allow the program to keep running after uncaught exception. Note that
      // we don't allow this function to be an async function, so that this
      // throw is truly synchronous.
      if (event === 'uncaughtException') {
        throw maybeError;
      }

      if (event.startsWith('SIG')) {
        // Wait for the child to finish. In a signal handler we can afford to
        // be asynchronous, unlike with 'exit' or 'uncaughtException'.
        childPromise.then(() => {
          // Re-signal self. Since this was a 'once', the signal won't hit this
          // handler again. Re-signaling ourself means that our exit status will
          // be as if we were killed by the signal we received, which is
          // something that things like nodemon like to see.
          process.kill(process.pid, event);
        });
      }
    };
    (process as EventEmitter).once(event, handler);
    return handler;
  }

  private setUpStopEvents(events: string[]) {
    const handlers = new Map<string, eventHandler>();
    events.forEach(event => {
      // Process any given event at most once.
      if (handlers.has(event)) {
        return;
      }
      handlers.set(event, this.setUpStopEvent(event));
    });
    this.eventStopper = () => {
      handlers.forEach((handler, event) => {
        process.removeListener(event, handler);
      });
      delete this.eventStopper;
    };
  }
}

// Literal IPv6 addresses contain colons and need to be wrapped in square
// brackets (like Go's net.JoinHostPort).
export function joinHostPort(host: string, port: number) {
  if (host.includes(':')) {
    host = `[${host}]`;
  }
  return `${host}:${port}`;
}
